<a href="https://colab.research.google.com/github/prisar/ai_notebooks/blob/main/nb_089.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Train a PPO or A2C agent from Stable Baselines3 to solve the LunarLander-v2 continuous control environment.

## Install dependencies

### Subtask:
Install the necessary libraries, including `stable-baselines3`, `huggingface-sb3`, `shimmy`, and `gymnasium`.


**Reasoning**:
The user wants to install the necessary libraries for the project. I will install all the required libraries in a single code block.



In [1]:
!pip install "stable-baselines3[extra]" huggingface-sb3 shimmy gymnasium

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 363.4/363.4 MB 3.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.8/13.8 MB 80.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.6/24.6 MB 71.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 883.7/883.7 kB 50.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 664.8/664.8 MB 1.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 211.5/211.5 MB 5.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 56.3/56.3 MB 12.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 127.9/127.9 MB 7.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 207.5/207.5 MB 5.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 21.1/21.1 MB 40.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 187.2/187.2 kB 17.0 MB/s eta 0:00:00
  Attempting uninstall: nvidia-nvjitlink-cu12
    Found existing installation: nvidia-nvjitlink-cu12 12.5.82
    Uninstall

## Import libraries

### Subtask:
Import the necessary modules from the installed libraries.


**Reasoning**:
I will import the necessary modules as instructed in a single code block. `gymnasium` will be imported as `gym`, `PPO` will be imported from `stable_baselines3`, and `evaluate_policy` will be imported from `stable_baselines3.common.evaluation`.



In [2]:
import gymnasium as gym
from stable_baselines3 import PPO
from stable_baselines3.common.evaluation import evaluate_policy

## Create the environment

### Subtask:
Create the LunarLander-v2 environment.


**Reasoning**:
I will now create the LunarLander-v2 environment with continuous actions using `gym.make()`, store it in the `env` variable, and reset it to get the initial observation, as per the instructions.



In [3]:
env = gym.make("LunarLander-v2", continuous=True)
obs, info = env.reset()

/usr/local/lib/python3.11/dist-packages/gymnasium/envs/registration.py:512: DeprecationWarning: WARN: The environment LunarLander-v2 is out of date. You should consider upgrading to version `v3`.
  logger.deprecation(


DeprecatedEnv: Environment version v2 for `LunarLander` is deprecated. Please use `LunarLander-v3` instead.

**Reasoning**:
The previous attempt to create the environment failed because version `v2` of `LunarLander` is deprecated. I will now correct this by using the recommended version `v3` and re-run the code to create and reset the environment.



In [4]:
env = gym.make("LunarLander-v3", continuous=True)
obs, info = env.reset()

DependencyNotInstalled: Box2D is not installed, you can install it by run `pip install swig` followed by `pip install "gymnasium[box2d]"`

**Reasoning**:
The previous attempt to create the environment failed because the `Box2D` dependency is not installed. I will now install `swig` and `gymnasium[box2d]` as suggested by the error message.



In [5]:
!pip install swig
!pip install "gymnasium[box2d]"

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 31.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 374.4/374.4 kB 10.3 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for box2d-py: filename=box2d_py-2.3.5-cp311-cp311-linux_x86_64.whl size=2379410 sha256=8e6cd5e7f7d43c97b4dce1248960e723164be71d42f1b60a38010df0b657d696
  Stored in directory: /root/.cache/pip/wheels/ab/f1/0c/d56f4a2bdd12bae0a0693ec33f2f0daadb5eb9753c78fa5308
Successfully built box2d-py


**Reasoning**:
Now that the `Box2D` dependency is installed, I can re-run the code to create and reset the `LunarLander-v3` environment.



In [6]:
env = gym.make("LunarLander-v3", continuous=True)
obs, info = env.reset()

<frozen importlib._bootstrap>:241: DeprecationWarning: builtin type SwigPyPacked has no __module__ attribute
<frozen importlib._bootstrap>:241: DeprecationWarning: builtin type SwigPyObject has no __module__ attribute
<frozen importlib._bootstrap>:241: DeprecationWarning: builtin type swigvarlink has no __module__ attribute
/usr/local/lib/python3.11/dist-packages/pygame/pkgdata.py:25: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
  from pkg_resources import resource_stream, resource_exists
/usr/local/lib/python3.11/dist-packages/pkg_resources/__init__.py:3154: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
  declare_namespace(pkg)
/usr/local/lib/python3.11/dist-packages/

## Instantiate the agent

### Subtask:
Instantiate a PPO agent from Stable Baselines3.


**Reasoning**:
I will now instantiate the PPO agent with the specified parameters, as per the instructions.



In [7]:
model = PPO(
    policy="MlpPolicy",
    env=env,
    verbose=1
)

Using cuda device
Wrapping the env with a `Monitor` wrapper
Wrapping the env in a DummyVecEnv.


/usr/local/lib/python3.11/dist-packages/stable_baselines3/common/on_policy_algorithm.py:150: UserWarning: You are trying to run PPO on the GPU, but it is primarily intended to run on the CPU when not using a CNN policy (you are using ActorCriticPolicy which should be a MlpPolicy). See https://github.com/DLR-RM/stable-baselines3/issues/1245 for more info. You can pass `device='cpu'` or `export CUDA_VISIBLE_DEVICES=` to force using the CPU.Note: The model will train, but the GPU utilization will be poor and the training might take longer than on CPU.
  warnings.warn(


## Train the agent

### Subtask:
Train the PPO agent on the LunarLander-v2 environment for a total of 50,000 timesteps.


**Reasoning**:
The next step is to train the PPO agent on the LunarLander-v2 environment for 50,000 timesteps, as specified in the instructions. I will call the `learn` method on the `model` object, setting `total_timesteps` to 50,000 and `progress_bar` to `True` to visualize the training process.



In [8]:
model.learn(total_timesteps=50000, progress_bar=True)

Output()

/usr/local/lib/python3.11/dist-packages/ipywidgets/widgets/widget_output.py:111: DeprecationWarning: 
Kernel._parent_header is deprecated in ipykernel 6. Use .get_parent()
  if ip and hasattr(ip, 'kernel') and hasattr(ip.kernel, '_parent_header'):

---------------------------------
| rollout/           |          |
|    ep_len_mean     | 120      |
|    ep_rew_mean     | -298     |
| time/              |          |
|    fps             | 568      |
|    iterations      | 1        |
|    time_elapsed    | 3        |
|    total_timesteps | 2048     |
---------------------------------
------------------------------------------
| rollout/                |              |
|    ep_len_mean          | 118          |
|    ep_rew_mean          | -281         |
| time/                   |              |
|    fps                  | 479          |
|    iterations           | 2            |
|    time_elapsed         | 8            |
|    total_timesteps      | 4096         |
| train/                  |              |
|    approx_kl            | 0.0045628757 |
|    clip_fraction        | 0.0334       |
|    clip_range           | 0.2          |
|    entropy_loss         | -2.85        |
|    explained_variance   | 0.00668      |
|    learning_r

## Evaluate the agent

### Subtask:
Evaluate the trained agent's performance.


**Reasoning**:
To evaluate the trained agent's performance, I will use the `evaluate_policy` function from `stable_baselines3.common.evaluation`. This function runs the agent in the environment for a specified number of episodes and returns the mean and standard deviation of the rewards. I will set the number of evaluation episodes to 10 as requested, store the results in `mean_reward` and `std_reward`, and then print these values to the console.



In [9]:
mean_reward, std_reward = evaluate_policy(model, env, n_eval_episodes=10)

print(f"Mean reward: {mean_reward:.2f} +/- {std_reward:.2f}")

/usr/local/lib/python3.11/dist-packages/stable_baselines3/common/evaluation.py:70: UserWarning: Evaluation environment is not wrapped with a ``Monitor`` wrapper. This may result in reporting modified episode lengths and rewards, if other wrappers happen to modify these. Consider wrapping environment first with ``Monitor`` wrapper.
  warnings.warn(


Mean reward: 63.45 +/- 163.80


## Summary:

### Data Analysis Key Findings
*   A PPO agent was successfully trained on the `LunarLander-v2` environment for 50,000 timesteps.
*   During training, the agent showed significant improvement, with the mean episode reward increasing from -298 to -70.2.
*   Post-training evaluation over 10 episodes yielded a mean reward of 63.45 with a standard deviation of 163.80.

### Insights or Next Steps
*   The trained agent has learned a decent policy but has not yet solved the environment, as a score of 200+ is typically required. To improve performance, consider increasing the `total_timesteps` for training.
*   The high standard deviation in the evaluation results suggests that the agent's performance is inconsistent. Further hyperparameter tuning or a longer training duration could help stabilize its performance.
